File: Residential_controller.py
class Column():
    def __init__(self, floors, cages):
        self.cages = cages
        self.floors = floors
        self.cagesList = []
        self.requestButtonsList = []
        for i in range(cages):
            self.cagesList.append(Cage(0, floors))
        for i in range(floors):
            if i == 0:
                self.requestButtonsList.append(RequestButton(i, "up"))
            else:
                self.requestButtonsList.append(RequestButton(i, "up"))
                self.requestButtonsList.append(RequestButton(i, "down"))

    def findCage(self, requestedFloor, direction):

        bestGap = self.floors
        chosenCage = None

        for cage in self.cagesList:
            if cage.direction == "up" and direction == "up" and requestedFloor < cage.currentFloor:
                chosenCage = cage
            elif cage.direction == "down" and direction == "down" and requestedFloor > cage.currentFloor:
                chosenCage = cage
            elif cage.status == "idle":
                chosenCage = cage
            else:
                for cage in self.cagesList:
                    gap = abs(cage.currentFloor - requestedFloor)
                if gap < bestGap:
                        chosenCage = cage
                        bestGap = gap
        print("Optimal cage found on floor " + str(chosenCage.currentFloor))
        return chosenCage

    def requestFloor(self, cage, requestedFloor):
        print("Sending cage from floor" + str(cage.currentFloor) + " to floor " + str(requestedFloor))
        cage.addToQueue(requestedFloor)
        cage.closeDoors()
        cage.move()

    def requestCage(self, requestedFloor, direction):
        print("Called cage to floor " + str(requestedFloor))

        cage = self.findCage(requestedFloor, direction)

        cage.addToQueue(requestedFloor)
        cage.move()
        return cage

class Cage():
    def __init__(self, currentFloor, floors):
        self.direction =  None
        self.floors = floors
        self.currentFloor = currentFloor
        self.status = "idle"
        self.queue = []
        self.cageButtonsList = []
        self.door = "closed"

        for i in range(self.floors):
            self.cageButtonsList.append(CageButton(i))

    def addToQueue(self, requestedFloor):
        self.queue.append(requestedFloor)

        if self.direction == "down":
            self.queue.sort(reverse=True)
        if self.direction == "up":
            self.queue.sort(reverse=True)


        print("Added floor " + str(requestedFloor) + " to the cage's queue. Current queue: " + ', '.join(str(x) for x in self.queue))

    def move(self):
        print ("Moving cage")
        while len(self.queue) > 0:

            firstElement = self.queue[0]

            if self.door == "open":
                print("Waiting for doorway to be cleared")
                self.closeDoors()
            if firstElement == self.currentFloor:
                del self.queue[0]
                self.openDoors()
            elif firstElement > self.currentFloor:
                self.status = "moving"
                self.direction = "up"
                self.moveUp()
            elif firstElement < self.currentFloor:
                self.status = "moving"
                self.direction = "down"
                self.moveDown()
        print("Waiting for door to be cleared")
        self.closeDoors()
        print("Cage is now idle")
        self.status = "idle"

    def moveDown(self):
        self.currentFloor -= 1
        print("Descending Cage currently on floor " + str(self.currentFloor))
	
    def moveUp(self):
        self.currentFloor += 1
        print("Ascending Cage currently on floor " + str(self.currentFloor))

    def closeDoors(self):
        self.door="closed"
        print("Closing doors")

    def openDoors(self):
        self.door = "open"
        print("Opening doors")



class RequestButton():
    def __init__(self, requestFloor, direction):
        self.requestFloor = requestFloor
        self.direction = direction

class CageButton():
    def __init__(self, floor):
        self.floor = floor


print("->Test<-")

def Test1():
    column1 = Column(10, 2)

    column1.cagesList[0].currentFloor = 2
    column1.cagesList[0].direction  =  ""
    column1.cagesList[0].status =  "idle"
    column1.cagesList[0].queue = []

    column1.cagesList[1].currentFloor = 6
    column1.cagesList[1].direction  =  ""
    column1.cagesList[1].status =  "idle"
    column1.cagesList[1].queue = []

    column1.requestCage(3, "up")

Test1()

File: Residential_controller.js
class Column {
    constructor(floors, cages) {
        this.floors = floors;
        this.cages = cages;
        this.cagesList = [];
        this.requestButtonList = [];
        for (let i = 0; i < cages; i++) {
            this.cagesList.push(new Cage(1, floors));
        }
		for (let i = 0; i < this.floors; i++) {
            if (i === 0) {
                this.requestButtonList.push(new RequestButton(i, "up", false));
                
            }else if (i === 9) {
                this.requestButtonList.push(new RequestButton(i, "down", false));
            }else {
                this.requestButtonList.push(new RequestButton(i, "up", false));
                this.requestButtonList.push(new RequestButton(i, "down", false));
            }
		}
    }

    findCage(requestedFloor, direction) {       //finding the appropriate elevator by the following circumstances://

		var bestGap = this.floors;
		var chosenCage = null;

		for (let i = 0; i < this.cagesList.length; i++) { //if elevator and request are ascending and the request is above the elevator
			if (this.cagesList[i].direction === "up" && direction === "up" && requestedFloor > this.cagesList[i].currentFloor) {
                chosenCage = this.cagesList[i];
			}else if (this.cagesList[i].direction === "down" && direction === "down" && requestedFloor < this.cagesList[i].currentFloor) {
                chosenCage = this.cagesList[i]; 
                     //if elevator and request are descending and the request is below the elevator
            }else if (this.cagesList[i].status == "idle") {
                chosenCage = this.cagesList[i];
                    //else if the status is idle
			}else {
				for (let i = 0; i < this.cagesList.length; i++) {
                    let gap = Math.abs(this.cagesList[i].currentFloor - requestedFloor); //absolute value of the elevator floor - requested floor
                    if (gap < bestGap) {
                        chosenCage = this.cagesList[i];
                        bestGap = gap;   //shortest gap becomes the chosen gap
                    }
                }
			}
		}
		console.log("Optimal cage found on floor " + chosenCage.currentFloor);
		return chosenCage; //console.log output displaying the chosen elevator
	}
	requestCage(requestedFloor, direction) {

		console.log("Called cage to floor " + requestedFloor);
		
		let cage = this.findCage(requestedFloor, direction);

		cage.addToQueue(requestedFloor);
        cage.move();
		return cage;
	}

	requestFloor(cage, requestedFloor) {
        console.log("Sending cage from floor " + cage.currentFloor + " to floor " + requestedFloor);
		cage.addToQueue(requestedFloor);
		cage.closeDoors();
		cage.move();
    }
}

class Cage { 
	constructor(currentFloor, floors) {

		this.direction = null;
		this.floors = floors;
		this.currentFloor = currentFloor;
		this.status = "idle";
		this.queue = [];
		this.cageButtonList = [];
		this.door = "closed";

		for (let i = 0; i < this.floors; i++) {
            this.cageButtonList.push(new CageButton(i, false));
        }
    }
	addToQueue(requestedFloor) {
		this.queue.push(requestedFloor)

		if (this.direction == "up") {
			this.queue.sort((a, b) => a - b)
		}
		else if (this.direction == "down") {
			this.queue.sort((a, b) => b - a)
		}

		console.log("Added floor " + requestedFloor + " to the cage's queue. Current queue: " + this.queue.join(", "));
	}
	move() {
		console.log("Moving cage");
		while (this.queue.length > 0) {

            let firstElement = this.queue[0];
            
			if (firstElement == this.currentFloor) {
				this.queue.shift();
				this.openDoors();
			}
			if (firstElement > this.currentFloor) {
				this.status = "moving";
				this.direction = "up";
				this.moveUp();
			}
			if (firstElement < this.currentFloor) {
				this.status = "moving";
				this.direction = "down";
				this.moveDown()
			}
		}
		if (this.queue.length === 0) {
			console.log("Unloading Passengers");
			this.closeDoors();
			console.log("Cage is now idle");
			this.status = "idle";
		}
	}
	moveDown() {
		this.currentFloor--;
		console.log("descending cage currently on floor " + this.currentFloor);
    }
    
    moveUp() {
		this.currentFloor++;
		console.log("ascending cage currently on floor " + this.currentFloor);
    }
    
	openDoors() {
			this.door = "open"
			console.log("<> Opened doors");
	}
	
	closeDoors() {
			this.door="closed"
			console.log(">< Closed doors");
	}

}

class RequestButton {
	constructor(requestFloor, direction, buttonPressed) {
		this.requestFloor = requestFloor;
		this.direction = direction;
		this.pressed = buttonPressed;
	}
}

class CageButton {
	constructor(floor, buttonPressed) {
		this.floor = floor;
		this.buttonPressed = buttonPressed;
	}
}

console.log("->Test<-")

function Test1() {
	
	columntest = new Column(10, 2);

	columntest.cagesList[0].currentFloor = 2;
    columntest.cagesList[0].direction = "";
    columntest.cagesList[0].status = "idle";
	columntest.cagesList[0].queue = [];

	columntest.cagesList[1].currentFloor = 6;
	columntest.cagesList[1].direction = "";
	columntest.cagesList[1].status = "idle";
	columntest.cagesList[1].queue = [];

	columntest.requestCage(3, "up");
}

Test1();
